src/librustc/infer/error_reporting/util.rs

```diff
@@ -79,22 +106,26 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
 
     // This method returns whether the given Region is Anonymous
     // and returns the DefId and the BoundRegion corresponding to the given region.
+    // The is_anon_anon is set true when we are dealing with cases where
+    // both the regions are anonymous i.e. E0623.
     pub fn is_suitable_anonymous_region(&self,
-                                        region: Region<'tcx>)
-                                        -> Option<(DefId, ty::BoundRegion)> {
+                                        region: Region<'tcx>,
+                                        is_anon_anon: bool)
+                                        -> Option<FreeRegionInfo> {
         if let ty::ReFree(ref free_region) = *region {
-            if let ty::BrAnon(..) = free_region.bound_region{
-                    let anonymous_region_binding_scope = free_region.scope;
-                    let node_id = self.tcx
-                        .hir
-                        .as_local_node_id(anonymous_region_binding_scope)
-                        .unwrap();
-                    match self.tcx.hir.find(node_id) {
-                        Some(hir_map::NodeItem(..)) |
-                        Some(hir_map::NodeTraitItem(..)) => {
-                            // Success -- proceed to return Some below
-                        }
-                        Some(hir_map::NodeImplItem(..)) => {
+            if let ty::BrAnon(..) = free_region.bound_region {
+                let anonymous_region_binding_scope = free_region.scope;
```

Comment:
Actually, this bunch of code only detects whether we are in an impl item. Could you *extract* it to an `is_bound_region_in_impl_item` function and call it from whenever its needed?